### ISSUE TITLE
A Seat is marked as CanGraphical=yes too early (or with false positives)

### ISSUE THREAD

**Issue by 3v1n0** [systemd/systemd#10435]
**systemd version the issue has been seen with**
> 239

**Used distribution**
 > Ubuntu 18.10

**Expected behaviour you didn't see**
 > `CanGraphical` is true only when there's an actual graphical session

**Unexpected behaviour you saw**
 > `CanGraphical` is set to on as soon there's a `fb*` device although this might not allow a proper graphical session

**Steps to reproduce the problem**
 - Install ubuntu 18.10 in a kvm machine using `qxl` driver ((mine is configured as `kvm -drive if=virtio,file=disk.qcow2 -m 4096 -vga qxl`)
 - Start the session (this issue happens mostly when a fast SSD is there).
 - Gdm is unable to start because `CanGraphical` is set to `yes` too early, when the `vesa-framebuffer` is available
 - Once the actual driver is loaded this goes suddenly back to `no`, and then again `yes` (see [upstream issue](https://gitlab.gnome.org/GNOME/gdm/issues/428))
 - This happens because `/dev/fb0` is found, but the relative `/dev/dri/card0` is not *yet* ready (so another failures, if we try to start gdm now)
 - The automatic internal mechanism gdm has to reload, tries to start again but fails because of various issues that are currently [handled in this pull request](https://gitlab.gnome.org/GNOME/gdm/merge_requests/37).

Here's a [journalctl.log](https://github.com/systemd/systemd/files/2487535/journalctl.log) with `udevadm` monitoring the changes in the background.

So, while gdm needs indeed to be stronger at handling such changes (but as said we're handling it), and indeed the driver could be improved, I'm not sure if systemd should just mark any presency of `/dev/fb*` as something that allows a graphical session. This is indeed true in the generic case (and using simple fb graphics), but maybe systemd should wait a bit more before advertising such seats as graphical, especially if a `/dev/dri/card*` can arrive soon.

I was wondering if having a more restrictive rule on setting the `master-of-seat` such as the one below could be a way, but indeed there might be many other cases to handle:

`SUBSYSTEM=="graphics", KERNEL=="fb[0-9]*", DRIVER!="vesa-framebuffer", TAG+="seat", TAG+="master-of-seat"`

Then maybe it could be instead advertised a graphical mode too, instead of just considering a framebuffer the same thing of a dri card.

Related ot this is also the fact that for the cases where the card is probed by X from PCI (vmware driver on kvm), `CanGraphical` never turns into `yes`, while X is perfectly able to run. Not sure if it's a driver issue though.

Feel free to mark this as not-fix as I can see the low-level role of systemd that can't do specific things here, while should be probably be gdm more restrictive, but I guess this is better to be discussed here than at upper levels of the stack.

**Comment by poettering** [comment]
Well, we never know what will show up. It might be dri without fb, and fb without dri, or both in combination. I figure it should be up to gdm/x11/wayland/gnome to figure out what to use there, and probably wait for the drm to actually show up before doing anything. I really don't think it's logind's job to do it...

The difference is really that gdm/x11/waylnad/gnome actually knows what it needs and want, and in the lower levels we can only guess.

hence, sorry, but I don't think this is something we can or should tweak, sorry!

**Comment by 3v1n0** [comment]
@poettering I see the point for that, however what if instead logind would not do some basic analysis on the `master-of-seat` device, in case that's a fb with dri?

I've prepared the pull #10452 to do this, it wouldn't look bad at my eyes, but not sure if I didn't consider something.

**Comment by poettering** [comment]
@3v1n0 but are there any guarantees that dri will always show up before fb? i don't think so. I mean, the in /sys both devices have no parent/child relationship, which means they can show up in any order they like. udev might process them in any order and in parallel (udev only makes order of processing guarantees regarding parent/child devices), and drivers are completely free to announce them in any order they like...

**Comment by 3v1n0** [comment]
@poettering I see that, and in such case this change won't do anything than keeping the current state of things.

it's also true that in case this happens (and in all the problematic causes which lead us here), this never happened.

So, while indeed we can't predict what will happen next, if the device is already there, but not announced by udev yet, we have a chance to fix it.

**Comment by poettering** [comment]
> So, while indeed we can't predict what will happen next, if the device is already there, but not announced by udev yet, we have a chance to fix it.

but how is this a good thing if it sometimes work and other times does not?

or can you point me to a somewhat official declaration somewhere that there's a guarantee that an fb and a drm device have to show up in a specific order?

**Comment by 3v1n0** [comment]
> but how is this a good thing if it sometimes work and other times does not?

I'd see instead the other side, how can be a bad thing if sometimes helps :joy:?

Thing is that while might not be announced here it seems to available for such check in all the cases I tried (mostly `i915`, `qxl` and `vboxvideo`), and thus in these scenarios when fb is announced we just wait to announce CanGraphical until the dri card is announced too.

> or can you point me to a somewhat official declaration somewhere that there's a guarantee that an fb and a drm device have to show up in a specific order?

Unfortunately there's not. But again if the dri card arrives earlier we just announce it, and thus there's no change in it.

**Comment by halfline** [comment]
> I figure it should be up to gdm/x11/wayland/gnome to figure out what to use there, and probably wait for the drm to actually show up before doing anything. I really don't think it's logind's job to do it...

Then what's the point of `CanGraphical` ?  I mean there's no advantage for GDM to listen for it right? As it stands, it flips to true when we shouldn't start a display server, and fails to flip to true in some cases when we should. If it's GDM's responsibility to figure things out anyway, then maybe `CanGraphical` should get deprecated.

**Comment by poettering** [comment]
> Then what's the point of `CanGraphical` ? I mean there's no advantage for GDM to listen for it right? As it stands, it sometimes flips to `true` when we shouldn't start a display server, and fails to flip to true in some cases when we should. If it's GDM's responsibility to figure things out anyway, then maybe `CanGraphical` should get deprecated.

Well, gdm is not the only theoretical user of logind. Moreover, `CanGraphical` doesn't lie, it will report true as soon as a graphical API is available, but it doesn't specify if that's the old framebuffer API or the new DRM API, and I am very sure it shouldn't report that.

Of course we could add CanDRM or so, which is more specific, but I am not sure it's worth the trouble, gdm/gnome/wayland/x11 might as well check the devices they need themselves there's very little point in second guessing that in logind, it can never know what precisely gdm/gnome/wayland/x11 will actually require from the set of devices...

My suggestion for gdm is to watch which seats get added, ignore those that report CanGraphical=false, and for those doing CanGraphical=true ignore them until they actually offer the devices gdm needs, i.e. a DRM device or so.

**Comment by halfline** [comment]
> Well, gdm is not the only theoretical user of logind. 

Sure, but you added `CanGraphical` for GDM with the intention of it fixing this problem: 

https://gitlab.gnome.org/GNOME/gdm/issues/103

:-)

> My suggestion for gdm is to watch which seats get added, ignore those that report CanGraphical=false, 

So the problem is X works in some cases with cards that don't have any kernel support, so they'll return CanGraphical=false atm.  Maybe fixable with the right kind of udev rule, not sure.

> and for those doing CanGraphical=true ignore them until they actually offer the devices gdm needs, i.e. a DRM device or so.

But I guess i could just skip the `CanGraphical` check entirely and just wait for the devices to show up?
I mean it's just extra code that doesn't actually buy us much right?

**Comment by 3v1n0** [comment]
> So the problem is X works in some cases with cards that don't have any kernel support, so they'll return `CanGraphical=false` atm. Maybe fixable with the right kind of udev rule, not sure.

Right, we do this for ati/nvidia stuff in ubuntu (https://github.com/tseliot/ubuntu-drivers-common/blob/master/share/hybrid/71-u-d-c-gpu-detection.rules), but this is not always the case (like in the case of vmware driver when used in kvm. In such case, we've also a quite weird behavior because after some waiting a card is added temporary and then removed again after few moments...).

But probably drivers that doesn't support a standard dri/fb interface as fallback might still define a `masster-of-seat` device when found with custom rule.

Anyway, the point of this issue/MP isn't defining when we've a a DRM card based experience, but just avoiding that we mark a `CanGraphical` for fb's linked to DRM cards. And this is something that has been done for years so far in playmouth (as per [this old @halfline change](https://gitlab.freedesktop.org/plymouth/plymouth/commit/c66c207d2e25ae77d118a6037db35454678a49dc))

**Comment by poettering** [comment]
So, hmm, what about this compromise: let's just ignore framebuffer devices? I mean, it's 2018, and noone wants to use them anyway, right? i.e. is there a logind-using display manager that would still care for framebuffer devices? If not, then maybe let's take the step and decide that DRM is the future and fb is not, and just deprecated it from our pov, under the assumption that code still using the framebuffer would in turn not bother with logind anyway?

**Comment by 3v1n0** [comment]
I didn't want to be so disruptive as there might still cases where wayland compositors run directly on FB's, but if you guys agree, I guess those cases could be handled by drivers by adding rules to mark devices as `master-of-seat`.

**Comment by halfline** [comment]
works for me.

**Comment by poettering** [comment]
Apparently this broke #11299.

**Comment by 3v1n0** [comment]
Mmhmh, so going back to my previous proposal is an option or what else?

**Comment by poettering** [comment]
I am tempted to say we should probably whitelist hyperv's fb and see. As long as the list of fb drivers to whitelist is short I figure we can get away with it.
